### PR TITLE
Implement postRemoved to all searchapi fixes #3110

### LINKED ISSUE #3110: Search API is not notified if we recycle a message
RemoveTopic.php

```
    // Only remove posts if they're not recycled.
    if (!$recycle)
    {
        // Callback for search APIs to do their thing
        require_once($sourcedir . '/Search.php');
        $searchAPI = findSearchAPI();
        if ($searchAPI->supportsMethod('postRemoved'))
            $searchAPI->postRemoved($message);
```

We only seem to notify the SearchAPI if we are not recycling a post.  This means the API will be out of sync until it manually corrects it.  This does not affect deleting an entire topic.

The way I see it we can fix this 2 ways.
1. We could call postRemoved anyways but with a second parameter called $recycle and let the API handle it if we are recycling a message/topic.
2. We call postModified method and let it treat it as if the post was modified.
3. Setup a new search api method to handle this.

### PR DESCRIPTION
**Pull request by albertlast**
When a post get changed we delete all related search results

Issue: #3110

### CODE REVIEW

**Comment by albertlast** [comment]
Since the coding was generic,
i saw no reason to keep this for fulltext only -> define this as default method

File: Sources/Class-SearchAPI.php [hunk 1]
```diff
@@ -202,6 +220,48 @@ public function postModified(array &$msgOptions, array &$topicOptions, array &$p
...
+		$id_searchs = array();
+		while ($row = $smcFunc['db_fetch_assoc']($result))
+			$id_searchs[] = $row['id_search'];
+
+		if (count($id_searchs) < 1)
+			return;
+
+		$smcFunc['db_query']('','
+			DELETE FROM {db_prefix}log_search_results
+			WHERE id_search in ({array_int:id_searchs})',
```
Comment (Sesquipedalian) [inline Sources/Class-SearchAPI.php:244]: This would delete many rows for posts besides the desired post.

**Review by Sesquipedalian — COMMENTED** [review]
The `id_search` column is not a unique key, so removing all rows that match the `id_search` will delete a lot of things that should not be deleted.

File: Sources/Class-SearchAPI.php (hunk 1 above)
Comment (albertlast) [reply Sources/Class-SearchAPI.php:244]: I don't want to delete only the desire post,
i want to "mark" the search cache als staled and
this archive i with the complete delete of the cache for this search.

**Comment by Sesquipedalian** [comment]
For example, running this in my database:
```
SELECT DISTINCT id_search
FROM dev_log_search_results
WHERE id_msg = 1;
```
gives this result:
```
1
4
6
10
12
14
16
18
```

If I were to run this:
```
DELETE FROM dev_log_search_results
WHERE id_search in (
	SELECT DISTINCT id_search
	FROM dev_log_search_results
	WHERE id_msg = 1
)
```
It would delete 278 rows in my dev_log_search_results table, most of which are not about msg 1 at all.

**Comment by albertlast** [comment]
Like i writen down in the uper post,
i want to delete all searches completly where msg 1 appears.
So the number of ~280 rows are okay in my eyes.

**Comment by Sesquipedalian** [comment]
The query as it is now would delete a lot of rows where msg 1 does not not appear.

When I run this query in my database:
```
SELECT * 
FROM dev_log_search_results 
WHERE id_search IN (
	SELECT DISTINCT id_search
	FROM dev_log_search_results
	WHERE id_msg = 1
);
```

I get this result:

id_search|id_topic|id_msg|relevance|num_matches
---------|--------|------|---------|-----------
1|1|1|550|1
1|2|2|550|1
1|3|14|555|6
1|4|6|550|1
1|5|7|550|1
1|6|8|550|1
1|7|9|550|1
1|8|12|550|1
1|9|16|551|2
1|10|15|550|1
1|11|17|550|1
1|14|20|550|1
1|15|21|550|1
1|16|25|553|4
1|29|26|550|1
1|30|29|551|2
1|31|33|551|2
1|32|53|552|3
1|33|31|550|1
1|49|49|550|1
1|51|51|550|1
1|52|52|550|1
1|53|161|795|44
1|54|109|556|7
1|55|99|551|2
1|68|120|555|1
1|72|124|574|1
1|73|125|579|1
1|107|160|747|1
1|108|168|788|3
1|109|163|762|1
1|111|167|781|1
1|112|169|790|1
1|113|171|801|2
4|1|1|450|1
4|2|2|450|1
4|3|3|1955|1
4|4|6|450|1
4|5|7|450|1
4|6|8|450|1
4|7|9|450|1
4|8|12|450|1
4|9|13|751|1
4|10|15|450|1
4|11|17|450|1
4|14|20|450|1
4|15|21|450|1
4|16|22|1353|1
4|29|26|450|1
4|30|27|751|1
4|31|28|751|1
4|32|30|1052|1
4|33|31|450|1
4|49|49|450|1
4|51|51|450|1
4|52|52|450|1
4|53|54|13393|1
4|54|96|2256|1
4|55|98|751|1
4|68|120|455|1
4|72|124|474|1
4|73|125|479|1
4|107|160|647|1
4|108|162|1259|1
4|109|163|662|1
4|111|167|681|1
4|112|169|690|1
4|113|170|996|1
6|1|1|550|1
6|2|2|550|1
6|3|14|555|6
6|4|6|550|1
6|5|7|550|1
6|6|8|550|1
6|7|9|550|1
6|8|12|550|1
6|9|16|551|2
6|10|15|550|1
6|11|17|550|1
6|14|20|550|1
6|15|21|550|1
6|16|25|553|4
6|29|26|550|1
6|30|29|551|2
6|31|33|551|2
6|32|53|552|3
6|33|31|550|1
6|49|49|550|1
6|51|51|550|1
6|52|52|550|1
6|53|161|790|44
6|54|109|556|7
6|55|99|551|2
6|68|120|550|1
6|72|124|569|1
6|73|125|574|1
6|107|160|742|1
6|108|168|783|3
6|109|163|757|1
6|111|167|776|1
6|112|169|786|1
6|113|171|796|2
6|114|172|800|1
10|1|1|550|1
10|2|2|550|1
10|3|14|555|6
10|4|6|550|1
10|5|7|550|1
10|6|8|550|1
10|7|9|550|1
10|8|12|550|1
10|9|16|551|2
10|10|15|550|1
10|11|17|550|1
10|14|20|550|1
10|15|21|550|1
10|16|25|553|4
10|29|26|550|1
10|30|29|551|2
10|31|33|551|2
10|32|53|552|3
10|33|31|550|1
10|49|49|550|1
10|51|51|550|1
10|52|52|550|1
10|53|161|790|44
10|54|109|556|7
10|55|99|551|2
10|68|120|550|1
10|72|124|569|1
10|73|125|574|1
10|107|160|742|1
10|108|168|783|3
10|109|163|757|1
10|111|167|776|1
10|112|169|786|1
10|113|171|796|2
10|114|172|800|1
12|1|1|550|1
12|2|2|550|1
12|3|14|555|6
12|4|6|550|1
12|5|7|550|1
12|6|8|550|1
12|7|9|550|1
12|8|12|550|1
12|9|16|551|2
12|10|15|550|1
12|11|17|550|1
12|14|20|550|1
12|15|21|550|1
12|16|25|553|4
12|29|26|550|1
12|30|29|551|2
12|31|33|551|2
12|32|53|552|3
12|33|31|550|1
12|49|49|550|1
12|51|51|550|1
12|52|52|550|1
12|53|161|790|44
12|54|109|556|7
12|55|99|551|2
12|68|120|550|1
12|72|124|569|1
12|73|125|574|1
12|107|160|742|1
12|108|168|783|3
12|109|163|757|1
12|111|167|776|1
12|112|169|786|1
12|113|171|796|2
12|114|172|800|1
14|1|1|450|1
14|2|2|450|1
14|3|3|1955|1
14|4|6|450|1
14|5|7|450|1
14|6|8|450|1
14|7|9|450|1
14|8|12|450|1
14|9|13|751|1
14|10|15|450|1
14|11|17|450|1
14|14|20|450|1
14|15|21|450|1
14|16|22|1353|1
14|29|26|450|1
14|30|27|751|1
14|31|28|751|1
14|32|30|1052|1
14|33|31|450|1
14|49|49|450|1
14|51|51|450|1
14|52|52|450|1
14|53|54|13393|1
14|54|96|2256|1
14|55|98|751|1
14|68|120|450|1
14|72|124|469|1
14|73|125|474|1
14|107|160|642|1
14|108|162|1254|1
14|109|163|657|1
14|111|167|676|1
14|112|169|686|1
14|113|170|991|1
14|114|172|700|1
16|1|1|450|1
16|2|2|450|1
16|3|3|1955|1
16|4|6|450|1
16|5|7|450|1
16|6|8|450|1
16|7|9|450|1
16|8|12|450|1
16|9|13|751|1
16|10|15|450|1
16|11|17|450|1
16|14|20|450|1
16|15|21|450|1
16|16|22|1353|1
16|29|26|450|1
16|30|27|751|1
16|31|28|751|1
16|32|30|1052|1
16|33|31|450|1
16|49|49|450|1
16|51|51|450|1
16|52|52|450|1
16|53|54|13393|1
16|54|96|2256|1
16|55|98|751|1
16|68|120|450|1
16|72|124|469|1
16|73|125|474|1
16|107|160|642|1
16|108|162|1254|1
16|109|163|657|1
16|111|167|676|1
16|112|169|686|1
16|113|170|991|1
16|114|172|700|1
18|1|1|450|1
18|2|2|450|1
18|3|3|1955|1
18|4|6|450|1
18|5|7|450|1
18|6|8|450|1
18|7|9|450|1
18|8|12|450|1
18|9|13|751|1
18|10|15|450|1
18|11|17|450|1
18|14|20|450|1
18|15|21|450|1
18|16|22|1353|1
18|29|26|450|1
18|30|27|751|1
18|31|28|751|1
18|32|30|1052|1
18|33|31|450|1
18|49|49|450|1
18|51|51|450|1
18|52|52|450|1
18|53|54|13393|1
18|54|96|2256|1
18|55|98|751|1
18|68|120|450|1
18|72|124|469|1
18|73|125|474|1
18|107|160|642|1
18|108|162|1254|1
18|109|163|657|1
18|111|167|676|1
18|112|169|686|1
18|113|170|991|1
18|114|172|700|1

If I had run `DELETE` instead of `SELECT *`, that would have deleted a lot of rows that have nothing to do with msg 1.

**Comment by albertlast** [comment]
Like you table show,
they all related with msg 1,
so the result is fine.

The relationship is -> msg 1 appears in:
```
1
4
6
10
12
14
16
18
```

To mark all this search cache results as stale i had to delete them completly.

**Comment by Sesquipedalian** [comment]
Oh, yes, that does make sense. Clearly it is too late here, and I should go to sleep rather than trying to review code right now. 😄 I'll come back to this later.

**Comment by albertlast** [comment]
what is the status or this pr @Sesquipedalian ?

**Comment by Sesquipedalian** [comment]
Its status is unchanged. My status is that I am currently travelling for work related purposes, so it is unlikely that I will be able to change its status until next week at the earliest. 🙂

**Review by colinschoen — APPROVED** [review]
LGTM

**Pull request merged**